### PR TITLE
Ft automated extraction joint

### LINKED ISSUE #57: Automated extraction of joint ids
Joint indices for reading in the limits, controlling the joints, and disable castor wheels is hard-coded at the moment. 
It would be much better if this was done automatically, or at least by name.

A starting point could be:
https://github.com/maxspahn/gym_envs_urdf/blob/d4d49dc1bb7486f5ac006efaf79e8ce08603d7c9/urdfenvs/tiagoReacher/resources/tiagoRobot.py#L23-L53

### PR DESCRIPTION
**Pull request by SkyloveQiu**
The first version is finished, resolve #57.
There are a few problems, to make the extraction of id, I introduce preload idea which means loading the robot before generating the environment. 
I have noticed that the preload will slow down the start-up process.

### CODE REVIEW

**Comment by SkyloveQiu** [comment]
The code needs to clean up, this pr just for testing purposes.

File: urdfenvs/tiago_reacher/resources/tiago_robot.py [hunk 1]
```diff
@@ -14,49 +14,31 @@ def __init__(self):
...
+        for i, joint in enumerate(robot.joints): 
+            if joint.name in self._joint_names: 
+                self._urdf_joints.append(i) 
+        self._robot_joints = [] 
+        self._castor_joints = [] 
+        for index in joint_info:
+            if joint_info[index].decode("UTF-8") in self._joint_names:
+                self._robot_joints.append(index)
+            if "caster" in joint_info[index].decode("UTF-8"):
+                self._castor_joints.append(index)
```
Comment (maxspahn) [inline urdfenvs/tiago_reacher/resources/tiago_robot.py:41]: That doesn't look too bad. I think you understood the idea behind it.
I will comment a bit more in the comment on this PR.

File: urdfenvs/urdfCommon/differential_drive_robot.py [hunk 2]
```diff
@@ -99,7 +109,7 @@ def get_observation_space(self) -> gym.spaces.Dict:
         `velocity` the concatenated velocities of joints in their local configuration space.
         `forward_velocity` the forward velocity of the robot.
         """
-
+        self.preload()
```
Comment (maxspahn) [inline urdfenvs/urdfCommon/differential_drive_robot.py:112]: The preloading does not relate to the observation space so it should not be called inside this particular function but rather in either the constructor or the reset function.

**Review by maxspahn — COMMENTED** [review]
Great initiative on this issue! It is not a simple one.

The goal of this PR is to replace hard-coded joint indices by hard-coded joint names. 
That is more intuitive and less prone to errors caused by changes in the urdf-file. 
The idea is to set the actuated joint names for every robot, similar to what you did in the tiago-robot. Then there should be a extractor function in the generic robot probably that takes these joint names and maps them to the correct indices. These indices can then be processed as done previously. 

Let me know if you have additional questions.

**Comment by SkyloveQiu** [comment]
preload has been removed. I changed the order of `set_space`, I moved it to the end of reset()

**Comment by SkyloveQiu** [comment]
How do I classify the joints? currently I use name to classify it, is there anyway to automate this process?

**Comment by SkyloveQiu** [comment]
I have updated the function `get_indexed_joint_info` in `genetic_robot`, now the function will use `_joint_names` to update `_robot_joints` and `_castor_joints`.

**Comment by SkyloveQiu** [comment]
the bug is observed, other robots have different start-up procedures compared to Tiago robots

**Comment by maxspahn** [comment]
You can also test your changes locally using `pytest`

**Comment by SkyloveQiu** [comment]
The bug is found, for point robot, it adds a lidar sensor after, in the previous design that reset will call `set_space` function which will remove lidar sensor space which causes an error. The pytest passed with warnings on my machine but it may different on CI.

File: examples/point_robot.py [hunk 3]
```diff
@@ -36,6 +36,7 @@ def main():
         env.add_goal(splineGoal)
     for _ in range(n_steps):
         ob, _, _, _ = env.step(action)
+        print(ob)
```
Comment (maxspahn) [inline examples/point_robot.py:39]: Can you remove all debug-print statements?

File: examples/tiago.py [hunk 4]
```diff
@@ -24,6 +24,7 @@ def main():
     vel0 = np.zeros(19)
     # vel0[0:2] = np.array([0.0, 0.0])
     # vel0[5:12] = np.array([0.1, 0.1, 0.2, -0.1, 0.1, 0.2, 0.0])
+    print("resetting")
```
Comment (maxspahn) [inline examples/tiago.py:27]: same here

**Review by maxspahn — CHANGES_REQUESTED** [review]
I will verify it on my system.
Can you still remove all the debug printout statements?

**Comment by SkyloveQiu** [comment]
I am not sure how to do the Prius robot, it seems not the same as other robots.

**Comment by SkyloveQiu** [comment]
it seems one pytest fails, not sure why.

**Comment by SkyloveQiu** [comment]
All other robots have been updated, and the pytest passed on my machine.

File: urdfenvs/albert_reacher/resources/albert_robot.py [hunk 5]
```diff
@@ -13,9 +13,18 @@ def __init__(self):
...
+        self._joint_names = ( 
+            wheel_joint_names 
+            + mmrobot_joint_name 
+        ) 
+        robot = URDF.load(self._urdf_file) 
+        self._urdf_joints = [] 
+        for i, joint in enumerate(robot.joints): 
+            if joint.name in self._joint_names: 
+                self._urdf_joints.append(i) 
+        self.get_indexed_joint_info()
```
Comment (maxspahn) [inline urdfenvs/albert_reacher/resources/albert_robot.py:27]: Can't we move that to a central function? Currently, these lines are in all the robot files.

File: urdfenvs/albert_reacher/resources/albert_robot.py [hunk 6]
```diff
@@ -13,9 +13,18 @@ def __init__(self):
         self._wheel_distance = 0.494
 
     def set_joint_indices(self):
```
Comment (maxspahn) [inline urdfenvs/albert_reacher/resources/albert_robot.py:15]: After your changes, the name of this function needs to be modified to something like `set_joint_names`

File: urdfenvs/urdfCommon/generic_robot.py [hunk 7]
```diff
@@ -65,18 +62,24 @@ def read_limits(self) -> None:
     def set_acceleration_limits(self):
         pass
 
-    def get_indexed_joint_info(self) -> dict:
+    def get_indexed_joint_info(self) -> None:
         """Get indexed joint info.
 
         This function can be used for debugging and finding
         the correct joint indices for self.setJointIndices.
```
Comment (maxspahn) [inline urdfenvs/urdfCommon/generic_robot.py:69]: This function is now central, so the comment on debugging and finding the correct joint indices should be modified.

File: urdfenvs/urdfCommon/generic_robot.py [hunk 8]
```diff
@@ -65,18 +62,24 @@ def read_limits(self) -> None:
...
             joint_info = p.getJointInfo(self._robot, i)
-            indexed_joint_info[joint_info[0]] = joint_info[1]
-        return indexed_joint_info
+            joint_name = joint_info[1].decode("UTF-8")
+            print(joint_name)
+            if joint_name in self._joint_names:
+                self._robot_joints.append(i)
+            if "castor" in joint_name:
+                self._castor_joints.append(i)
+
```
Comment (maxspahn) [inline urdfenvs/urdfCommon/generic_robot.py:82]: Concerning the other comment on repeated code lines: This is where I expect self._urdf_joints to be set, just like you do it with the other joint indices.

**Review by maxspahn — CHANGES_REQUESTED** [review]
Can we get rid of the last tiny bit of duplicated code, concerning `self._urdf_joints`

File: urdfenvs/albert_reacher/resources/albert_robot.py [hunk 9]
```diff
@@ -12,10 +11,14 @@ def __init__(self):
...
-        self._robot_joints = [24, 25, 8, 9, 10, 11, 12, 13, 14]
-        self._castor_joints = [22, 23]
+    def set_joint_names(self):
+        wheel_joint_names = ["wheel_right_joint", "wheel_left_joint"] 
+        mmrobot_joint_name = ["mmrobot_joint" + str(i) for i in range(1,8)] 
+        self._joint_names = ( 
+            wheel_joint_names 
+            + mmrobot_joint_name 
+        )
+        self.get_indexed_joint_info()
```
Comment (maxspahn) [inline urdfenvs/albert_reacher/resources/albert_robot.py:21]: This function should be called from the generic robot class.

**Review by maxspahn — CHANGES_REQUESTED** [review]
Only a minor change to respect.
One function -> one purpose.

**Review by maxspahn — APPROVED** [review]
Great work.

**Comment by maxspahn** [comment]
Can you reopen this pull request. I needed to revert it, as there were too many errors which I did not check carefully enough.
Will comment on them in the new PR.

**Comment by SkyloveQiu** [comment]
The revert is not processed correctly, it shows nothing to compare in my site.

**Comment by maxspahn** [comment]
You cannot reopen a merged pull-request. I created a new pull request reverting this one.
Now, you should open a new PR from this branch.

**Pull request merged**